Add tests for accnames from hidden labels, captions, and legends

adampage opened the pull request:
Accname tests to support w3c/html-aam#533’s clarification that hidden `label`s, `caption`s, and `legend`s will not provide an accname to their naturally associated elements unless they are explicitly referenced using `aria-labelledby`.

adampage committed:
add tests for names from hidden elements

adampage committed:
normalize class names

cookiecrook reviewed (approved):
Looks okay to me, but not merging immediately since we only got one CI run.

scottaohara commented:
would this PR be the place to also create the tests for how browsers should use other naming mechanisms if these elements are hidden?

e.g.,

```
<label for=d hidden>
  this should not be used, cause it'll return the empty string
</label>
<input id=d title="this should be used">
```

adampage committed:
Merge remote-tracking branch 'upstream/master' into html-aam-hidden-names

adampage committed:
remove tentative test status

adampage committed:
add accname tests by element

adampage committed:
remove iframe src in case it’s causing CI Firefox crash

adampage commented:
> would this PR be the place to also create the tests for how browsers should use other naming mechanisms if these elements are hidden?

So... I took ☝🏻 this ball and ran with it... maybe in the wrong direction. 😵‍💫

I focused on HTML-AAM’s “[**4.1** Accname computation by element](https://www.w3.org/TR/html-aam-1.0/#accname-computation)” section and wrote/copypasta’d a boatload of new tests for each specific element identified there.

Every test specimen follows the same pattern: I simultaneously throw all possible accname techniques at the element, then I assert which one should _win_ based on the conditionals described in the spec. Next, I eliminate 1 technique at a time for subsequent test specimens, rinse and repeat until the element has no remaining accname candidates.

I took a few notes along the way for a few tweaks I might propose in the HTML-AAM spec. I also left 2 sections incomplete:

- [**4.1.7** Other form elements](https://www.w3.org/TR/html-aam-1.0/#other-form-elements-accessible-name-computation) — I just did `input type="radio"`, `input type="checkbox"`, and `select` as a start.
- [**4.1.16** Section and grouping elements](https://www.w3.org/TR/html-aam-1.0/#section-and-grouping-element-accessible-name-computation) — I just did `section`, `article`, and `nav` as a start.

Anyhow, I’ll pause here. [The test results](https://wpt.fyi/results/html-aam/accname-computation-by-element?sha=b70094e&label=pr_head) are interesting. Please let me know what you think, @scottaohara / @cookiecrook.

cookiecrook reviewed (dismissed):
Updating because this PR increased in scope since my last review. I shared offline feedback with Adam: mainly that the numbered directories might be fragile (as the spec changes) and that they may complicate the browsing experience on wpt.fyi. He also plans to utilize a new JavaScript convenience method to reduce some of the markup duplication.

adampage committed:
Merge remote-tracking branch 'upstream/master' into html-aam-hidden-names

adampage committed:
replaces all static subtest markup with dynamic generation

adampage committed:
Merge remote-tracking branch 'upstream/master' into html-aam-hidden-names

adampage commented:
Alrightey, @cookiecrook and @scottaohara, I just committed a monster of an update. 😬 

This change:

- Removes the [brittle numbered directory names](https://wpt.fyi/results/html-aam/accname-computation-by-element?sha=b70094e&label=pr_head) corresponding to the HTML-AAM computation sections.
- Replaces the static subtests with [dynamic markup generation](https://github.com/web-platform-tests/wpt/pull/44965/files#diff-de7b88e6abae2f964cd620997ad7a41d3027fc9e65745964bd2da4b9f3a940b9).

Now that the subtests are being built dynamically, I was able to quickly introduce a bunch more name-supporting HTML elements that I‘d neglected in my previous push (e.g., `<thead>`, `<progress>`, `<hr>` 🤯 , `<input type="color">`, etc.).

As a result, the number of subtests have almost doubled from [276](https://wpt.fyi/results/html-aam/accname-computation-by-element?sha=b70094e&label=pr_head) to [510](https://wpt.fyi/results/html-aam/accname-computation-by-element?label=pr_head&max-count=1&pr=44965).

I also scattered a few `TODO` notes around, such as [this bit about phrasing elements](https://github.com/web-platform-tests/wpt/pull/44965/files#diff-e79e5be0f633ff34b53eae30d8c3cfa847e48626de1f35a9d858cce928ce68dfR31-R32) which I think dovetails with ARIA issues you’ve filed recently, @scottaohara / @rahimabdi. Many of these notes will require deeper investigation on my part. I’m hoping most of them will be resolved either by:

1. Proposing updates to the HTML-AAM spec itself
2. Updating/removing specific test expectations
3. Accepting a failed test result and filing browser bugs

There are also a number of tests that are failing because of web driver errors, [such as `<optgroup>`](https://wpt.fyi/results/html-aam/accname-computation-by-element/other-form-elements.html?label=pr_head&max-count=1&pr=44965). I’m not sure what do with those, if anything?

At the end of all of this, I’m looking at HTML-AAM’s [Accessible Name Computations By HTML Element](https://www.w3.org/TR/html-aam-1.0/#accname-computation) section and thinking it could be simplified. Most of the computation steps follow a very linear path through a prioritized set of potential naming sources. My gut says we can define a very small number of “rule sets” and then make a table to pair each nameable HTML element with one of those rule sets. cc: @rahimabdi

cookiecrook commented on accname/name/comp_hidden_not_referenced.html:
```diff
@@ -73,6 +73,36 @@
   </span>
 </h2>
 
+<h2>Input with hidden label</h2>
+<label id="hidden-label" for="input-01" hidden>Hidden label</label>
+<input
+  id="input-01"
+  class="ex-label"
+  data-testname="Input that is aria-labelledby hidden label"
+  data-expectedlabel="Hidden label"
+  aria-labelledby="hidden-label"
+>
+
+<h2>Table with hidden caption</h2>
+<table
+  class="ex-label"
+  data-testname="Table that is aria-labelledby hidden caption"
+  data-expectedlabel="Hidden caption"
+  aria-labelledby="hidden-caption"
+>
+  <caption id="hidden-caption" hidden>Hidden caption</caption>
+</table>
+
+<h2>Fieldset with hidden legend</h2>
+<fieldset
+  class="ex-label"
+  data-testname="Fieldset that is aria-labelledby hidden legend"
+  data-expectedlabel="Hidden legend"
+  aria-labelledby="hidden-legend"
+>
+  <legend id="hidden-legend" hidden>Hidden legend</legend>
+</fieldset>
+
```
New tests to a file tracked in the current Interop project would only be accepted if they didn't introduce new failures mid-year. In this case, the 3 new tests are all passing in Chromium, Gecko, and WebKit.... (Edge unknown but likely passing?) so I think it's okay to include this in the PR... With the caveat that we could roll it back in the unlikely chance it causes a failure for Edge.

In the future, it may be easier to include Interop-tracked changes as a separate PR from ones that do not affect the current Interop project.

adampage replied:
Ah, understood, thanks. I’ll make that distinction for future PRs and check in with you if I have doubts.

cookiecrook replied:
In cases where you need to add a new valid-but-failing tests, add a new file that uses the same file name +".tentative" and have it only include the new failing tests... e.g. `comp_hidden_not_referenced.tentative.html`

adampage replied:
Got it. 👍🏻  I did just manually confirm that latest Edge _is_ passing all 3 new tests, so... no harm no foul? But I’ll stay ready to roll it back regardless.

scottaohara replied:
going to mark this as resolved since these all still should be passing (and do when actually tested in a browser and viewing the a11y tree that is computed).

cookiecrook commented on html-aam/scripts/html-aam-accname-utils.js:
```diff
@@ -0,0 +1,338 @@
+
+
+  testId: 0,
+  srcImgSample: "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjgiIGhlaWdodD0iMzYiIGZpbGw9Im5vbmUiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+CiAgPHBhdGggZD0iTTIwIDZoLTJ2NGgtMlY2aC00djJoLTJ2MmgydjJIOFY4SDZ2NGgydjRoMTR2LTJoLTZ2LTJoMnYtMmgydjJoNnYtMmgtMlY4aC00VjZaIiBmaWxsPSIjRUZDMjk4Ii8+CiAgPHBhdGggZD0iTTYgNmg2djJoLTJ2MmgydjJIOFY4SDZWNloiIGZpbGw9IiMwMDAiLz4KICA8cGF0aCBkPSJNNiA4SDR2Nmg0di0ySDZWOFpt ...
```
This "Mario Bros" graphic probably has some copyright restrictions by Nintendo. It's also an SVG, which triggers different rendering pipelines than standard raster images.

FYI... there's a shorter, single-pixel transparent GIF in other tests:
`data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==`

adampage replied:
So much for my easter egg, `<audio src="sad-trombone-sound.mp3">`.

I should have known better. I’ll use that GIF.

adampage replied:
[Done](https://github.com/web-platform-tests/wpt/pull/44965/commits/9260cfb2bd48a23e973bbadb71a08391e0fbee74).

cookiecrook reviewed (commented):
incomplete review, and I plan to come back to it, but wanted to share feedback as I can since my work time is limited.

cookiecrook commented on html-aam/accname-computation-by-element/tr-td-th.html:
```diff
@@ -0,0 +1,48 @@
+          How about thead, tbody, and tfoot, which support name from author?
+        */
+
+        'thead': { parents: ['table'] },
+        'tbody': { parents: ['table'] },
+        'tfoot': { parents: ['table'] },
```
I'm not certain these tests would be valid w/o the relevant descendants, would they? IIRC, engines invalidate some containers if other required context is missing or invalid, so a `table` that contained no cell children would not be a table, and therefore your downstream `thead` expectations may be invalid.

Note: I'm reading this first one as testing `<table><thead></thead></table>` so if I'm missing some cell additions later, please correct.

adampage replied:
Oof, I didn’t even consider that. I’ve filled out the table structures for [all the `table`-related subtests](https://github.com/web-platform-tests/wpt/pull/44965/commits/80544531d21b0b90ef68b98610224b574b2cd2a6).

I did a quick once-over for the other similarly non-trivial markup structures — `area`, `ruby`, `summary`, etc. — and I believe those are all sufficiently complete.

adampage replied:
Now that’s interesting. After implementing your suggestion here, [Chrome has 2 new failures and Firefox has 1](https://wpt.fyi/results/html-aam/accname-computation-by-element/tr-td-th.html?label=pr_head&max-count=1&pr=44965).

With the previous (incomplete) subtests, [everything passed in flying colors](https://wpt.fyi/results/html-aam/accname-computation-by-element/tr-td-th.html?sha=84f1c0d&label=pr_head).

With the `table` elements now fully fleshed out, Chrome is happy to let both `th` and `td` get their names from their contents, which [the HTML-AAM spec doesn’t seem to permit](https://www.w3.org/TR/html-aam-1.0/#tr-td-th-elements-accessible-name-computation). This is also true for Firefox, but only for `th`.

cookiecrook replied:
I would then recommend marking this file (or just those tests in a new file) as tentative. Then spawn a new ARIA issue for the WG to determine what the "correct" behavior should be.

adampage replied:
Thanks, @cookiecrook. I’ve [marked that file as tentative](https://github.com/web-platform-tests/wpt/pull/44965/commits/ea6ca4cee9da1b02e5547c2ea1c988a135ee7729).

I also found that issue appears to exist already in w3c/html-aam#543, so I left [a comment](https://github.com/w3c/html-aam/issues/543#issuecomment-2415714902) to mention this test.

cookiecrook commented on html-aam/accname-computation-by-element/tr-td-th.html:
```diff
@@ -0,0 +1,48 @@
+    const testConfig = {
+      urlSpec: 'https://www.w3.org/TR/html-aam-1.0/#tr-td-th-elements-accessible-name-computation',
+      elements: {
+        'tr': { parents: ['table', 'tbody'] },
+        'th': { parents: ['table', 'thead', 'tr'] },
+        'td': { parents: ['table', 'tbody', 'tr'] },
```
Seems like the `parents` parameter should be `ancestors`? Is there a way to add any attributes to the ancestors?

adampage replied:
> Seems like the parents parameter should be ancestors?

Sigh, yes, that would have been much more accurate. 🤦🏻‍♂️  I’ve [renamed](https://github.com/web-platform-tests/wpt/pull/44965/commits/05325043d4cce5731fb2792136c95df0449e484c) `parents` to `ancestors` and `children` to `descendants`.

> Is there a way to add any attributes to the ancestors?

Not currently. I did consider it along the way, possibly using [selector parsing](https://github.com/web-platform-tests/wpt/pull/44965/files#diff-de7b88e6abae2f964cd620997ad7a41d3027fc9e65745964bd2da4b9f3a940b9R287-R291) like I’m already doing with the dictionary key for the test element itself. 🤔

cookiecrook commented on html-aam/scripts/html-aam-accname-utils.js:
```diff
@@ -0,0 +1,338 @@
+const HtmlAamAccnameUtils = {
```
Thanks for making this super configurable!

adampage replied:
Sure thing! Am hoping for maximum coverage with minimal maintenance. 🤞🏻

cookiecrook commented on html-aam/scripts/html-aam-accname-utils.js:
```diff
@@ -0,0 +1,338 @@
+      ]
+    }
+    HtmlAamAccnameUtils.buildNameComputationTests(testConfig);
+
```
After a shorter example source (e.g. "from title" only), it may be good to show the rendered DOM element(s) that will be generated from that source. E.g. `img` with parent `a href` using 'from title`

PS. can you use `a href` in the parent params, or would it need to be an unlinked anchor? Same question on `img` with an `input type=button` parent.

adampage replied:
If I’m understanding correctly, this might be pushing my scope even further beyond the [original intent](https://github.com/web-platform-tests/wpt/pull/44965#issue-2172762598)? 😅

It’s [my own damn fault](https://github.com/web-platform-tests/wpt/pull/44965#issuecomment-2063146364), really, but — at least for polishing off this PR — I’ve mainly been focused on generating “simple” subtests for the relatively straightforward conditions specified in [HTML-AAM 4.1](https://www.w3.org/TR/html-aam-1.0/#accessible-name-and-description-computation), particularly in terms of the _absence of potential naming sources_.  Just fleshing out those alone created 500+ subtests. 🤯  And I’m quite certain I’ve even missed some HTML elements as it is.

> PS. can you use a href in the parent params, or would it need to be an unlinked anchor? Same question on img with an input type=button parent.

I’m afraid not, currently — [same answer as above](https://github.com/web-platform-tests/wpt/pull/44965#discussion_r1623038114). But I’d like to add that! I’d like to go even more deeply into interesting subtree-sensitive permutations like `a[href][title] > img`, but maybe that could be a 2.0?

adampage replied:
Thanks, @cookiecrook. I’ve [marked that file as tentative](https://github.com/web-platform-tests/wpt/pull/44965/commits/ea6ca4cee9da1b02e5547c2ea1c988a135ee7729).

I also found that issue appears to exist already in w3c/html-aam#543, so I left [a comment](https://github.com/w3c/html-aam/issues/543#issuecomment-2415714902) to mention this test.

adampage committed:
replace sample image

adampage committed:
flesh out interior table content

adampage committed:
rename children/parents to descendants/ancestors

cookiecrook reviewed (commented):
follow-up comments interspersed

adampage committed:
render an unnamed image for invisible empty subtests

adampage committed:
make tr/td/th test tentative

scottaohara commented on html-aam/hidden-names.html:
```diff
@@ -0,0 +1,52 @@
+<body>
+
+  <h1>Accname tests for hidden elements that would otherwise provide names to an associated element</h1>
+  <p>These tests are a complement to <a href="https://github.com/w3c/html-aam/pull/533">w3c/html-aam#533</a>.</p>
```
re-reading the PR, i'm worried that one of the intended updates was not clear, as i don't see it represented in these tests (unless i missed it - i'm just reviewing the tests via github diff right now)

but for the hidden caption/legend examples, i would have liked there to be another instance where there was a hidden instance, and then a second instance that should then name the table/fieldset.

e.g.,

```
<table>
  <caption hidden>not the name</caption>
  <caption>is the name</caption>
  <!-- rest of the table stuff goes here -->
</table>
```

right now, i wouldn't be surprised if this fails in some/all browsers, so per @cookiecrook's prior comment, maybe this needs to just be a second PR.

adampage replied:
Thanks, @scottohara. I’ve added [two new subtests](https://github.com/web-platform-tests/wpt/pull/44965/commits/a520a8ab3cd0b169607aa6efe5af6056082f83c5) for that case in the static `hidden-names.html` test file. As you suspected, [the browser results are interesting](https://wpt.fyi/results/html-aam/hidden-names.html?label=pr_head&max-count=1&pr=44965).

In my auto-generated subtests for `table`/`caption` and `fieldset`/`legend`, I also test HTML-AAM’s clause about using [“_the subtree of the first such element_”](https://www.w3.org/TR/html-aam-1.0/#accname-computation:~:text=caption%20element%2C%20then-,use%20the%20subtree%20of%20the%20first%20such%20element,-.). All 3 browsers do pass [for both the `fieldset`](https://wpt.fyi/results/html-aam/accname-computation-by-element/fieldset.html?label=pr_head&max-count=1&pr=44965#:~:text=fieldset%20expecting%20name%20from%20first%20legend) and [the `table`](https://wpt.fyi/results/html-aam/accname-computation-by-element/table.html?label=pr_head&max-count=1&pr=44965#:~:text=table%20expecting%20name%20from%20first%20caption).

adampage committed:
add subtests for hidden+visible caption/legend

adampage committed:
Merge remote-tracking branch 'upstream/master' into html-aam-hidden-names

giacomo-petri commented on html-aam/accname-computation-by-element/figure.html:
```diff
@@ -0,0 +1,37 @@
+        'from 2 aria-labelledby refs',
+        'from 1 aria-labelledby ref',
+        'from aria-label',
+        'from first figcaption',
```
Based on this PR https://github.com/w3c/aria/pull/2224/files, figcaption is no longer accurate.

scottaohara replied:
as giacomo said above, figcaption is not meant to be a name source for figure, so let's remove this
```suggestion
```

giacomo-petri commented on html-aam/accname-computation-by-element/input-image.html:
```diff
@@ -0,0 +1,59 @@
+        'from 2 labels',
+        'from 1 label',
+        'from alt',
+        'from empty alt',
```
The specs don't clearly state whether alt="" should be used before title. Unlike with <img>, it's not explicitly defined how these attributes should impact if present together. In practice, most major browsers prioritize the title attribute over alt="".

scottaohara replied:
>If the [accessible name](https://www.w3.org/TR/accname-1.2/#dfn-accessible-name) is still empty and there is no alt attribute, use the title attribute.

step 3 in the img accname in html aam.  the intent there is that `title` provides the name only if there is _no `alt` at all_.

giacomo-petri commented on html-aam/accname-computation-by-element/inputs-with-placeholders.html:
```diff
@@ -0,0 +1,47 @@
+        'from 2 labels',
+        'from 1 label',
+        'from title',
+        'from placeholder',
```
It's not in the specs, but should we consider adding aria-placeholder between title and placeholder?

scottaohara replied:
seems the reality is that the native attribute takes precedent over the aria-placeholder attribute.

firefox on macOS is the only one i found that behaves a bit different, and VoiceOver actually would combine the values of `placeholder=n` and `aria-placeholder=p` to be announced as "n p"

going to make an HTML AAM update for this to resolve the "it's not in the specs" part of your comment :)

giacomo-petri commented on html-aam/accname-computation-by-element/section-and-grouping-elements.html:
```diff
@@ -0,0 +1,62 @@
+        'search': {},
+        'form': {},
+        'main': {},
+        'section': {},
```
Interestingly, a `<section>` without an accessible name is treated as a generic container, so "to be empty" might not apply in that context. That said, it may not be worth creating two separate test cases.

scottaohara replied:
fwiw, probably makes sense to leave as is.  since this is more about checking for names than what roles are exposed - which a few of these would have variant roles based on position/naming in the dom

scottaohara commented on html-aam/accname-computation-by-element/img.html:
```diff
@@ -0,0 +1,38 @@
+        'from alt',
+        'from empty alt',
+        'from title',
+        'to be empty'
```
per prior comment, we _can_ add in now that an `img` is SUPPOSED to get a name from a `figcaption` in the following markup example:

```
<figure>
  <img src=...> <!-- notice no alt -->
  <figcaption>i'm the accname for the image!</figcaption>
</figure>
```

though, this doesn't neatly align to the script, so not exactly sure how to recommend this be added, unless in a different file?

scottaohara commented on html-aam/accname-computation-by-element/inputs-with-placeholders.html:
```diff
@@ -0,0 +1,47 @@
+        'from 2 labels',
+        'from 1 label',
+        'from title',
+        'from placeholder',
```
since that last comment, [HTML AAM was updated](https://w3c.github.io/html-aam/#input-type-text-input-type-password-input-type-number-input-type-search-input-type-tel-input-type-email-input-type-url-and-textarea-elements-accessible-name-computation) to note that `placeholder` takes priority over `aria-placeholder.  so i'd imagine this would need to be called out like:

```suggestion
        'from placeholder',
        'from aria-placeholder',
```

scottaohara reviewed (commented):
did a quick review of this with @rahimabdi today.  some new comments for you @adampage  (for when you have time to look, again)

adampage commented:
Awesome, thanks very much, @scottaohara and @rahimabdi.  I’m keen to land this one. I’ll make time within the next couple weeks to review and update.

adampage committed:
Merge remote-tracking branch 'upstream/master' into html-aam-hidden-names

adampage committed:
figure is no longer named by figcaption

adampage committed:
add img named by ancestor figure figcaption

adampage committed:
add name from aria-placeholder

adampage commented:
Heya @scottaohara / @rahimabdi / @giacomo-petri / @cookiecrook, thank you for your reviews, and apologies for taking so long to get back this. I’ve reviewed the last year’s (🤯) worth of feedback and made these 3 updates:

1. Removed `figcaption` as a name source for `figure` — [latest results](https://wpt.fyi/results/html-aam/accname-computation-by-element/figure.html?label=pr_head&max-count=1&pr=44965)
2. Added `figcaption as a name source for an `img` element with a `figure` ancestor — [only Safari passes](https://wpt.fyi/results/html-aam/accname-computation-by-element/img.html?label=pr_head&max-count=1&pr=44965#:~:text=img%20expecting%20name%20from%20ancestor%20figure%20figcaption)
3. Added `aria-placeholder` as a name source after `placeholder` — [only Firefox fails](https://wpt.fyi/results/html-aam/accname-computation-by-element/inputs-with-placeholders.html?label=pr_head&max-count=1&pr=44965#:~:text=input%5Btype%3D%22email%22%5D%20expecting%20name%20from%20aria%2Dplaceholder)

Here’s the [latest browser results for the complete set](https://wpt.fyi/results/html-aam/accname-computation-by-element?label=pr_head&max-count=1&pr=44965).

rahimabdi commented:
LGTM, thank you very much @adampage!

I especially like the `html-aam/scripts/html-aam-accname-utils.js` utils which should make future testing easier.

giacomo-petri commented on html-aam/accname-computation-by-element/other-form-elements.html:
```diff
@@ -0,0 +1,81 @@
+        'input[type="date"]': {},
+        'input[type="datetime-local"]': {},
+        'input[type="file"]': {},
+        'input[type="hidden"]': {},
```
```suggestion
```

remove this line

adampage commented:
> For a potential follow-up PR, since we've added specific scenarios, it might be worth considering scenarios where the element is encapsulated by its label (e.g., a button or an input type button), in which case the label and its content/value should be excluded. e.g. button

Oh, nice one, @giacomo-petri — I hadn’t caught this update to HTML-AAM. I’ve added it to this PR via [`9953c9c`](https://github.com/web-platform-tests/wpt/pull/44965/commits/9953c9c3e877084ce3b6865a3f629d383e9829e4).

> **important** In other-form-elements.html, we should probably remove the input type="hidden", since it is removed from the accessibility tree, and I don't think we should be checking its acc name

Oh my goodness. 🤦🏻  Thank you for catching this. Fixed in [`96fea49`](https://github.com/web-platform-tests/wpt/pull/44965/commits/96fea49621b8c803f62b49cd29317f2d708fd2cb).

cookiecrook commented:
Merged due to 3 approving reviews... I did not personally re-review. Sorry for the delay @adampage.
